pkg/controller/ipam/antrea_ipam_controller.go

```diff
@@ -0,0 +1,243 @@
+type AntreaIPAMController struct {
+	// crdClient is the clientset for CRD API group.
+	crdClient versioned.Interface
+
+	// Pool cleanup events triggered by StatefulSet delete
+	statefulSetCleanupQueue workqueue.RateLimitingInterface
+	statefulSetMap          sync.Map
+
+	// follow changes for Namespace objects
+	namespaceLister       corelisters.NamespaceLister
+	namespaceListerSynced cache.InformerSynced
+
+	// follow changes for StatefulSet objects
+	statefulSetInformer     appsinformers.StatefulSetInformer
+	statefulSetListerSynced cache.InformerSynced
+
+	// follow changes for Pod objects
+	podLister       corelisters.PodLister
+	podListerSynced cache.InformerSynced
+
+	// follow changes for IP Pool objects
+	ipPoolLister       crdlisters.IPPoolLister
+	ipPoolListerSynced cache.InformerSynced
+}
+
+func NewAntreaIPAMController(crdClient versioned.Interface,
+	statefulSetInformer appsinformers.StatefulSetInformer,
+	namespaceInformer coreinformers.NamespaceInformer,
+	podInformer coreinformers.PodInformer,
+	ipPoolInformer crdinformers.IPPoolInformer) *AntreaIPAMController {
+	c := &AntreaIPAMController{
+		crdClient:               crdClient,
+		statefulSetCleanupQueue: workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "statefulSetCleanup"),
+		statefulSetMap:          sync.Map{},
+		namespaceLister:         namespaceInformer.Lister(),
+		namespaceListerSynced:   namespaceInformer.Informer().HasSynced,
+		podLister:               podInformer.Lister(),
+		podListerSynced:         podInformer.Informer().HasSynced,
+		statefulSetInformer:     statefulSetInformer,
+		statefulSetListerSynced: statefulSetInformer.Informer().HasSynced,
+		ipPoolLister:            ipPoolInformer.Lister(),
+		ipPoolListerSynced:      ipPoolInformer.Informer().HasSynced,
+	}
+
+	// Add handlers for Stateful Set events.
+	klog.V(2).InfoS("Subscribing for StatefulSet notifications", "controller", controllerName)
+	statefulSetInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			DeleteFunc: c.enqueueStatefulSetDeleteEvent,
+		},
+	)
+	return c
+}
+
+// Enqueue the STatefulSet delete notification to process by the worker
+func (c *AntreaIPAMController) enqueueStatefulSetDeleteEvent(obj interface{}) {
+	ss := obj.(*appsv1.StatefulSet)
+	klog.V(2).InfoS("Delete notification", "StatefulSet", ss.Name)
+
+	key := ss.GetUID()
+	c.statefulSetMap.Store(key, ss)
+	c.statefulSetCleanupQueue.Add(key)
+}
+
+// Find IP Pools annotated to StatefulSet via direct annotation or namespace annotation
+func (c *AntreaIPAMController) getIPPoolsByStatefulSet(ss *appsv1.StatefulSet) []string {
+
+	// Inspect pool annotation for the Pods
+	// In order to avoid extra API call in IPAM driver, IPAM annotations are defined
+	// on Pods rather than on StatefulSet
+	selector := map[string]string{"app": ss.Name}
+	podList, err := c.podLister.Pods(ss.Namespace).List(labels.SelectorFromSet(labels.Set(selector)))
```

Comment:
Why we need to query Pod to get the pool annotation? Doesn't the statefulset contain the pod template?

But from the perspective of cleanup, I think it's not comprehensive to rely on Pod, StatefulSet, or Namespace event/object, as the event could be missed if the controller is not running when the StatefulSet is removed. Instead, we could just add an statefulSetIndex to IPPool storage.

At startup, we compare whether there is any IP address allocated to non-existing StatefulSet, enqueue the StatefulSet if yes.
When receiving a deletion event of a StatefulSet, enqueue its Namespace+Name, instead of its UID as IPPool usage keeps Namespace+Name, and no need to store the StatefulSet in a map as we don't need other information.

In `processNextStatefulSetWorkItem`, get the IPPool that have IP allocated to the StatefulSet, then release the IP Addresses.

In this way, we don't need NamespaceInformer, NamespaceIndex, PodInformer, PodLister, etc., and can cover all cases.

Reply:
Agreed with Quan.
Here caused an issue https://github.com/antrea-io/antrea/pull/3141#issuecomment-1008479007

Reply:
Thanks for the idea, refactored.